# Project Recommender System

<table align="left">
    <tr>
        <td style="text-align:left">Kode Mata Kuliah</td>
        <td style="text-align:left">:</td>
        <td style="text-align:left">12S3205</td>
    </tr>
    <tr>
        <td style="text-align:left">Nama Mata Kuliah</td>
        <td style="text-align:left">:</td>
        <td style="text-align:left">Sistem Rekomendasi</td>
    </tr>
    <tr>
        <td style="text-align:left">Topik</td>
        <td style="text-align:left">:</td>
        <td style="text-align:left"><i>Sistem Rekomendasi untuk Pariwisata Toba</i></td>
    </tr>
    <tr>
        <td style="text-align:left">Mahasiswa 1</td>
        <td style="text-align:left">:</td>
        <td style="text-align:left"><i>12S21012 - Walker Valentinus Simanjuntak</i></td>
    </tr>
    <tr>
        <td style="text-align:left">Mahasiswa 2</td>
        <td style="text-align:left">:</td>
        <td style="text-align:left"><i>12S21021 - Naomi Elena Lumbanraja</i></td>
    </tr>
    <tr>
        <td style="text-align:left">Mahasiswa 3</td>
        <td style="text-align:left">:</td>
        <td style="text-align:left"><i>12S21033 - Sry Deviani Tambunan</i></td>
    </tr>
</table>

## Data Pre-processong

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [2]:
data = pd.read_csv("data/Tempat-Wisata-Toba-Preprocessing.csv")

In [3]:
data.shape

(43226, 9)

In [4]:
data.columns

Index(['Unnamed: 0.1', 'Unnamed: 0', 'address', 'PlaceID',
       'Nama_tempat_wisata', 'Category', 'ReviewerId', 'Rating', 'Reviews'],
      dtype='object')

In [5]:
data = data.rename(columns={'ReviewerId': 'user_id', 'PlaceID': 'item_id', 'Rating': 'rating'})
# data['user_id'] = data['user_id'].astype(str)
# data['item_id'] = data['item_id'].astype(str)
data['rating'] = data['rating'].astype(float)
data.dropna(inplace=True)
data.head(5)

Unnamed: 0.1  Unnamed: 0  \
1             1           1   
3             3           3   
4             4           4   
6             6           6   
8             8           8   

                                             address  item_id  \
1  Jl. Sibola Hotang, Sibola Hotangsas, Kec. Bali...        0   
3  Jl. Sibola Hotang, Sibola Hotangsas, Kec. Bali...        0   
4  Jl. Sibola Hotang, Sibola Hotangsas, Kec. Bali...        0   
6  Jl. Sibola Hotang, Sibola Hotangsas, Kec. Bali...        0   
8  Jl. Sibola Hotang, Sibola Hotangsas, Kec. Bali...        0   

          Nama_tempat_wisata       Category       user_id  rating  \
1  PASIR PUTIH LUMBAN BULBUL  Wisata Bahari  1.119090e+20     5.0   
3  PASIR PUTIH LUMBAN BULBUL  Wisata Bahari  1.130720e+20     5.0   
4  PASIR PUTIH LUMBAN BULBUL  Wisata Bahari  1.061730e+20     5.0   
6  PASIR PUTIH LUMBAN BULBUL  Wisata Bahari  1.142390e+20     5.0   
8  PASIR PUTIH LUMBAN BULBUL  Wisata Bahari  1.047430e+20     3.0   

                                             Reviews  
1                                              bagus  
3                                 sangat menyenagkan  
4                      bebas foto dimana aja cuma 2k  
6                         amazing pengen kesini lagi  
8  jalan menuju lokasi perlu diperhatikan oleh pe...

In [6]:
data.shape

(22166, 9)

In [7]:
# Step 1: Split into 85% training+validation and 15% test
train_val_data, test_data = train_test_split(data, test_size=0.15, random_state=42)
# Step 2: Define validation size as 17% of train+validation, rounded up
val_size = int(len(train_val_data) * 0.17) + 1
# Step 3: Split 85% data into training (about 70% of original) and validation (about 15% of original)
train_data, val_data = train_test_split(train_val_data, test_size=val_size, random_state=42)
# Check sizes
print(f"Train data size: {train_data.shape}")
print(f"Validation data size: {val_data.shape}")
print(f"Test data size: {test_data.shape}")

Train data size: (15638, 9)
Validation data size: (3203, 9)
Test data size: (3325, 9)


Penjelasan Langkah-Langkah

1. Membagi Data Menjadi Training+Validation dan Test:
   - train_test_split(data, test_size=0.15, random_state=42): Fungsi ini digunakan untuk membagi dataset awal menjadi dua bagian, yaitu:
     - 85% untuk Training+Validation (train_val_data)
     - 15% untuk Test (test_data)
   - Pembagian ini bertujuan untuk menyisihkan sebagian data sebagai data uji akhir (test data), yang akan digunakan untuk mengevaluasi kinerja model setelah tahap pelatihan dan tuning selesai.

2. Menentukan Ukuran Data Validation:
   - val_size = int(len(train_val_data) * 0.17) + 1:
     - Setelah membagi data awal menjadi training+validation dan test, kita menentukan bahwa 17% dari train_val_data akan digunakan sebagai data validasi.
     - Penambahan +1 memastikan bahwa ukuran validasi selalu dibulatkan ke atas, agar kita mendapatkan sekitar 17% dari
   - train_val_data sebagai data validasi.

3. Membagi Training+Validation Menjadi Training dan Validation:
   - train_test_split(train_val_data, test_size=val_size, random_state=42): Fungsi ini membagi train_val_data menjadi dua bagian:
     - Training Data (train_data), sekitar 70% dari data asli.
     - Validation Data (val_data), sekitar 15% dari data asli.
   - Pembagian ini bertujuan untuk menyediakan data validasi yang akan digunakan untuk tuning model, misalnya dalam memilih hyperparameter.

4. Menampilkan Ukuran Setiap Subset:
   - print(f"Train data size: {train_data.shape}"): Menampilkan ukuran data training.
   - print(f"Validation data size: {val_data.shape}"): Menampilkan ukuran data validasi.
   - print(f"Test data size: {test_data.shape}"): Menampilkan ukuran data uji.ilkan ukuran data uji.

Analisis Pembagian Data
1. Pembagian data yang dilakukan di sini cukup standar dalam machine learning dan sangat penting untuk memastikan bahwa model tidak overfitting atau underfitting.
2. Training Set (70%) digunakan untuk melatih model, mempelajari pola-pola dalam data, dan mengoptimalkan parameter model.
3. Validation Set (15%) menyediakan data yang terpisah dari training untuk tuning hyperparameter dan memilih model terbaik. Ini mencegah model dari overfitting pada training data karena kita melakukan evaluasi selama proses tuning pada data yang belum pernah dilihat oleh model.
4. Test Set (15%) adalah bagian dari data yang sepenuhnya terpisah, digunakan untuk mengevaluasi performa akhir model setelah proses pelatihan dan tuning selesai. Ini memberikan gambaran yang lebih objektif tentang bagaimana model akan berkinerja pada data baru atau data di dunia nyata.

## Model Pembanding CF - SVD

In [8]:
from sklearn.decomposition import TruncatedSVD
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_squared_error, mean_absolute_error
import matplotlib.pyplot as plt

In [9]:
def create_user_item_matrix(data, user_col='user_id', item_col='item_id', rating_col='rating'):
    data = data.groupby([user_col, item_col])[rating_col].mean().reset_index()
    return data.pivot(index=user_col, columns=item_col, values=rating_col).fillna(0)

# Apply the function to create matrices for each split
train_matrix = create_user_item_matrix(train_data)
val_matrix = create_user_item_matrix(val_data).reindex(columns=train_matrix.columns, fill_value=0)
test_matrix = create_user_item_matrix(test_data).reindex(columns=train_matrix.columns, fill_value=0)

In [10]:
train_matrix

item_id       0    1    3    4    5    6    8    9    10   11   ...  92   93   \
user_id                                                         ...             
1.000030e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000040e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000090e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000100e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000120e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
...           ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1.184390e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  5.0  0.0   
1.184410e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.184430e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  5.0   
1.184440e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.184460e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   

item_id       94   95   96   97   98   99   100  101  
user_id                                               
1.000030e+20  0.0  0.0  0.0  0.0  4.0  0.0  0.0  0.0  
1.000040e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.000090e+20  5.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.000100e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  5.0  
1.000120e+20  0.0  5.0  0.0  0.0  0.0  0.0  0.0  5.0  
...           ...  ...  ...  ...  ...  ...  ...  ...  
1.184390e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.184410e+20  0.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0  
1.184430e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.184440e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  5.0  
1.184460e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[8905 rows x 98 columns]

In [11]:
val_matrix

item_id       0    1    3    4    5    6    8    9    10   11   ...  92   93   \
user_id                                                         ...             
1.000090e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000140e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000160e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000230e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  5.0  0.0   
1.000270e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
...           ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1.184300e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.184330e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.184350e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  5.0  ...  0.0  0.0   
1.184430e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  5.0   
1.184440e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   

item_id       94   95   96   97   98   99   100  101  
user_id                                               
1.000090e+20  0.0  0.0  0.0  0.0  0.0  5.0  0.0  0.0  
1.000140e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.000160e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.000230e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.000270e+20  0.0  0.0  0.0  0.0  0.0  0.0  5.0  0.0  
...           ...  ...  ...  ...  ...  ...  ...  ...  
1.184300e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  5.0  
1.184330e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.184350e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.184430e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.184440e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[2736 rows x 98 columns]

In [12]:
test_matrix

item_id       0    1    3    4    5    6    8    9    10   11   ...  92   93   \
user_id                                                         ...             
1.000100e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000120e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000160e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000270e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.000340e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
...           ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1.184400e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.184410e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.184430e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.184440e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1.184460e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   

item_id       94   95   96   97   98   99   100  101  
user_id                                               
1.000100e+20  0.0  4.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.000120e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.000160e+20  0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  
1.000270e+20  0.0  0.0  0.0  0.0  5.0  0.0  0.0  0.0  
1.000340e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  5.0  
...           ...  ...  ...  ...  ...  ...  ...  ...  
1.184400e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.184410e+20  0.0  0.0  0.0  0.0  4.0  0.0  0.0  0.0  
1.184430e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.184440e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.184460e+20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[2838 rows x 98 columns]

Penjelasan Fungsi create_user_item_matrix
1. Parameter Fungsi:
   - data: DataFrame yang berisi informasi user_id, item_id, dan rating.
   - user_col: Nama kolom untuk ID pengguna, defaultnya 'user_id'.
   - item_col: Nama kolom untuk ID item, defaultnya 'item_id'.
   - rating_col: Nama kolom untuk rating, defaultnya 'rating'.

2. Langkah di dalam Fungsi:
   - data.groupby([user_col, item_col])[rating_col].mean().reset_index():
     - Langkah ini mengelompokkan data berdasarkan user_id dan item_id, lalu menghitung rata-rata rating dari setiap pasangan pengguna-item.
     - Jika ada beberapa rating untuk pasangan pengguna-item tertentu, maka nilai rating rata-rata akan digunakan.
   - data.pivot(index=user_col, columns=item_col, values=rating_col).fillna(0):
     - Fungsi pivot digunakan untuk mengonversi data menjadi format user-item matrix, dengan user_id sebagai indeks baris dan item_id sebagai kolom.
     - fillna(0): Mengisi nilai yang hilang dalam matriks dengan 0. Hal ini berarti bahwa jika pengguna belum memberikan rating untuk suatu item, maka nilai defaultnya adalah 0.

3. Mengaplikasikan Fungsi ke Set Data Training, Validation, dan Test:
   - train_matrix: Matriks user-item untuk data training.
   - val_matrix: Matriks user-item untuk data validation. Matriks ini disesuaikan agar kolomnya sama dengan train_matrix menggunakan reindex, dengan nilai default 0 untuk item yang mungkin tidak ada di data validasi tetapi ada di data training.
   - test_matrix: Matriks user-item untuk data test. Juga disesuaikan agar kolomnya sama dengan train_matrix dengan reindex.rix dengan reindex.

Analisis Kode
1. Manfaat Representasi User-Item Matrix:
   - Matriks ini memudahkan model rekomendasi, terutama model berbasis dekomposisi matriks (seperti SVD) untuk menemukan pola tersembunyi dalam data rating.
   - Dengan menyimpan data dalam format matriks, kita dapat mengaplikasikan algoritma dekomposisi matriks untuk menemukan faktor-faktor laten (latent factors) yang mendasari hubungan antara pengguna dan item.

2. Mengisi Nilai Kosong dengan 0:
   - Menggunakan fillna(0) memiliki makna bahwa pengguna yang belum menilai item tertentu dianggap memberikan rating 0. Ini bisa relevan dalam kasus tertentu, tetapi ada pendekatan lain yang mengisi nilai kosong dengan nilai rata-rata atau nilai lain yang lebih bermakna.
   - Pengisian dengan 0 juga membantu algoritma untuk berfokus pada rating yang nyata tanpa terpengaruh oleh nilai yang hilang.

3. Penggunaan reindex:
reindex memastikan bahwa val_matrix dan test_matrix memiliki struktur kolom yang sama dengan train_matrix. Ini penting agar model rekomendasi dapat bekerja dengan matriks yang konsisten dan memastikan bahwa semua item yang muncul di data training juga terwakili di data validation dan test, meskipun pengguna di validation dan test mungkin belum menilai item tersebut.m tersebut.

In [13]:
# Creating an SVD model
def svd_model(train_matrix, n_components=50):
    svd = TruncatedSVD(n_components=n_components, random_state=42)
    # Latih model SVD pada matriks latih
    svd.fit(train_matrix)
    return svd

In [14]:
# Train the SVD model
svd_model_trained = svd_model(train_matrix)

Analisis Kode
1. Manfaat Penggunaan SVD dalam Sistem Rekomendasi:
   - SVD membantu mendekomposisi matriks user-item yang besar dan jarang (sparse) menjadi ruang laten dengan dimensi yang lebih kecil, di mana setiap komponen laten menangkap hubungan tersembunyi antara pengguna dan item.
   - Model rekomendasi yang dihasilkan dapat memberikan saran yang lebih akurat, karena faktor laten ini membantu mengidentifikasi pola preferensi pengguna terhadap fitur-fitur umum dari item yang direkomendasikan.

2. Pemilihan n_components:
   - Pemilihan nilai n_components merupakan aspek penting dalam model SVD. Jumlah komponen laten yang optimal bergantung pada kompleksitas data dan kebutuhan model.
   - Nilai yang lebih rendah mungkin tidak menangkap cukup informasi, sementara nilai yang terlalu tinggi dapat menyebabkan model menjadi kompleks dan overfitting.
   - Dalam kasus ini, 50 komponen digunakan sebagai titik awal, dan nilai ini dapat disesuaikan atau diuji lebih lanjut untuk menemukan yang paling optimal berdasarkan performa model di data validasi.

3. Dekomposisi Matriks dan Ruang Laten:
   - Proses ini mengubah data dari matriks besar yang mungkin memiliki banyak nilai kosong menjadi representasi yang lebih padat dalam ruang laten, yang dapat digunakan untuk menemukan kemiripan antara pengguna dan item.
   - Setelah dekomposisi, model dapat merekomendasikan item dengan cara mengidentifikasi faktor-faktor laten yang memiliki keterkaitan dengan pengguna tertentu.

### Matrix Evaluation

In [15]:
# Project Test and Validation Matrix
def project_matrix(matrix, svd_model):
    matrix_reduced = svd_model.transform(matrix)
    
    matrix_approx = svd_model.inverse_transform(matrix_reduced)
    return matrix_approx

val_matrix_approx = project_matrix(val_matrix, svd_model_trained)
test_matrix_approx = project_matrix(test_matrix, svd_model_trained)

1. Proyeksi Matriks dengan SVD:
   - Tujuan dari proyeksi ini adalah untuk memperkirakan rating pada matriks validasi dan test menggunakan representasi laten yang ditemukan dari data training.
   - Dengan memproyeksikan ke ruang laten, model dapat menghasilkan perkiraan rating berdasarkan pola umum yang ditemukan dalam data latih.

2. Penggunaan transform dan inverse_transform:
   - transform mengonversi data ke dalam bentuk laten yang lebih kompak dan kaya informasi, memungkinkan kita untuk menangkap hubungan antar pengguna dan item dalam dimensi yang lebih rendah.
   - inverse_transform mengembalikan representasi ini ke bentuk aslinya dalam user-item matrix, tetapi dengan nilai yang diprediksi berdasarkan representasi laten tersebut.

3. Kegunaan val_matrix_approx dan test_matrix_approx:
   - Matriks val_matrix_approx dan test_matrix_approx yang dihasilkan akan memiliki perkiraan rating yang dapat digunakan untuk mengevaluasi model pada data validasi dan test.ang sesuai.

In [16]:
# Function to calculate RMSE
def compute_rmse(true_matrix, approx_matrix):
    return np.sqrt(mean_squared_error(true_matrix.ravel(), approx_matrix.ravel()))

# Functions for RMSE and MAE evaluation
def evaluate_rmse_mae(true_matrix, approx_matrix):
    # Convert to numpy array for evaluation
    true_matrix = true_matrix.values if isinstance(true_matrix, pd.DataFrame) else true_matrix
    approx_matrix = approx_matrix.values if isinstance(approx_matrix, pd.DataFrame) else approx_matrix
    rmse = compute_rmse(true_matrix, approx_matrix)
    mae = mean_absolute_error(true_matrix.ravel(), approx_matrix.ravel())
    return rmse, mae

# Function to project a matrix and return a DataFrame
def project_matrix_to_df(matrix, svd_model, columns):
    matrix_df = pd.DataFrame(matrix, columns=columns)
    matrix_reduced = svd_model.transform(matrix_df)
    matrix_approx = svd_model.inverse_transform(matrix_reduced)
    return pd.DataFrame(matrix_approx, columns=columns)

In [17]:
val_matrix_approx_df = project_matrix_to_df(val_matrix, svd_model_trained, val_matrix.columns)
test_matrix_approx_df = project_matrix_to_df(test_matrix, svd_model_trained, test_matrix.columns)

# Evaluation on validation and test matrices
val_rmse, val_mae = evaluate_rmse_mae(val_matrix, val_matrix_approx_df)
test_rmse, test_mae = evaluate_rmse_mae(test_matrix, test_matrix_approx_df)

In [18]:
# Output of evaluation results
print(f"Validation RMSE: {val_rmse}")
print(f"Validation MAE: {val_mae}")
print(f"Test RMSE: {test_rmse}")
print(f"Test MAE: {test_mae}")

Validation RMSE: 0.14595748681359338
Validation MAE: 0.016271202294922637
Test RMSE: 0.14311698431116252
Test MAE: 0.016289272287725712


Analisis Hasil
1. RMSE dan MAE yang Rendah:
   - RMSE maupun MAE pada data validasi dan test memiliki nilai yang sangat rendah. Ini menunjukkan bahwa prediksi model cukup akurat dalam mendekati nilai rating asli.
   - RMSE rendah menunjukkan bahwa rata-rata deviasi kuadrat dari prediksi terhadap rating asli kecil, sehingga model mampu meminimalkan kesalahan besar.
   - MAE rendah menunjukkan bahwa rata-rata kesalahan absolut (perbedaan langsung antara prediksi dan nilai asli) juga kecil, yang berarti bahwa model ini sangat baik dalam memprediksi nilai rating yang hampir tepat.

2. Konsistensi antara Validation dan Test:
   - Nilai RMSE dan MAE pada data validasi dan test sangat mirip, menunjukkan bahwa model memiliki generalization error yang rendah. Artinya, performa model pada data yang belum terlihat (data test) tidak jauh berbeda dengan performanya pada data validasi.
   - Ini menunjukkan bahwa model tidak mengalami overfitting pada data latih, sehingga model yang dihasilkan mampu bekerja dengan baik pada data baru.

In [19]:
def mean_average_precision_at_k(true_matrix, approx_matrix, k=5):
    map_score = 0
    total_users = true_matrix.shape[0]
    
    for user in range(total_users):
        true_ratings = true_matrix.iloc[user].values
        approx_ratings = approx_matrix.iloc[user].values
        
        # Get top-k recommendations based on predictions
        top_k_items = np.argsort(approx_ratings)[::-1][:k]
        relevant_items = np.argsort(true_ratings)[::-1][:k]  # item yang relevan berdasarkan rating asli
        
        # Calculate precision at each level k
        hits = len(set(top_k_items) & set(relevant_items))
        if hits > 0:
            map_score += hits / k
    
    return map_score / total_users if total_users > 0 else 0

In [20]:
# Calculate MAP on validation and test data
val_map = mean_average_precision_at_k(val_matrix, val_matrix_approx_df, k=5)
test_map = mean_average_precision_at_k(test_matrix, test_matrix_approx_df, k=5)

print(f"Validation MAP: {val_map}")
print(f"Test MAP: {test_map}")

Validation MAP: 0.2324561403508773
Test MAP: 0.23255813953488522


Analisis Hasil
1. Interpretasi Nilai MAP:
   - MAP pada data validasi dan test berada di sekitar 0.226. Ini menunjukkan bahwa, rata-rata, sekitar 22.6% dari rekomendasi top-5 yang diberikan oleh model adalah relevan (sesuai dengan item yang mendapat rating tinggi pada data asli).
   - Nilai MAP yang mendekati 1 menunjukkan performa yang sangat baik, sedangkan nilai yang lebih rendah menunjukkan bahwa model kurang akurat dalam memberikan rekomendasi item relevan di posisi teratas.

2. Konsistensi antara Validation dan Test:
   - Nilai MAP pada data validasi dan test cukup konsisten. Ini menunjukkan bahwa model tidak overfitting pada data latih dan mampu menjaga performanya pada data yang belum terlihat (data test).
   - Konsistensi ini juga menunjukkan bahwa model dapat memberikan rekomendasi yang cukup relevan pada data baru dengan tingkat presisi yang mirip seperti pada data validasi.


In [21]:
def precision_at_k(true_matrix, approx_matrix, k=5, threshold=4.0):
    precision_scores = []

    for user in range(true_matrix.shape[0]):
        # Original ratings and predictions for a particular user
        true_ratings = true_matrix.iloc[user].values
        approx_ratings = approx_matrix.iloc[user].values

        # Get top-k items with the highest predicted rating
        top_k_items = np.argsort(approx_ratings)[::-1][:k]
        
        # Count the number of relevant items in top-k predictions
        relevant_items = [i for i in top_k_items if true_ratings[i] >= threshold]
        precision = len(relevant_items) / k
        precision_scores.append(precision)
    
    return np.mean(precision_scores)

In [22]:
def recall_at_k(true_matrix, approx_matrix, k=5, threshold=4.0):
    recall_scores = []

    for user in range(true_matrix.shape[0]):
        # Original ratings and predictions for a particular user
        true_ratings = true_matrix.iloc[user].values
        approx_ratings = approx_matrix.iloc[user].values

        # Get the actual relevant items
        relevant_items = [i for i in range(len(true_ratings)) if true_ratings[i] >= threshold]
        if len(relevant_items) == 0:
            continue

        # Get top-k items with the highest predicted rating
        top_k_items = np.argsort(approx_ratings)[::-1][:k]
        
        # Count the number of relevant items in top-k predictions
        hits = len(set(top_k_items) & set(relevant_items))
        recall = hits / len(relevant_items)
        recall_scores.append(recall)
    
    return np.mean(recall_scores)

In [23]:
# Calculate Precision and Recall on validation data and test data
val_precision = precision_at_k(val_matrix, val_matrix_approx_df, k=5, threshold=4.0)
test_precision = precision_at_k(test_matrix, test_matrix_approx_df, k=5, threshold=4.0)

val_recall = recall_at_k(val_matrix, val_matrix_approx_df, k=5, threshold=4.0)
test_recall = recall_at_k(test_matrix, test_matrix_approx_df, k=5, threshold=4.0)

In [24]:
print(f"Validation Precision: {val_precision}")
print(f"Test Precision: {test_precision}")
print(f"Validation Recall: {val_recall}")
print(f"Test Recall: {test_recall}")

Validation Precision: 0.19568713450292402
Test Precision: 0.1952078928823115
Validation Recall: 0.9400161355385237
Test Recall: 0.9457455113192819


Analisis Hasil
1. Precision Rendah pada Top-5:
   - Nilai Precision pada data validasi dan test berada di sekitar 0.196, yang berarti bahwa sekitar 19.6% dari rekomendasi top-5 yang diberikan oleh model adalah item yang relevan (item yang sesuai dengan rating >= 4).
   - Precision rendah menunjukkan bahwa dari 5 rekomendasi teratas, hanya sedikit yang benar-benar sesuai dengan item yang disukai oleh pengguna. Ini menunjukkan bahwa model perlu ditingkatkan dalam merekomendasikan item yang relevan pada posisi teratas.

2. Recall Tinggi pada Top-5:
   - Nilai Recall sangat tinggi, yaitu sekitar 0.94 pada data validasi dan 0.947 pada data test.
   - Recall tinggi menunjukkan bahwa model berhasil menemukan hampir semua item relevan di antara semua item yang dianggap relevan oleh pengguna dalam top-5 rekomendasi.
   - Hal ini menunjukkan bahwa meskipun Precision rendah, model cukup baik dalam menemukan item relevan untuk pengguna, meskipun beberapa di antaranya mungkin tidak berada pada posisi teratas.

Precision yang rendah dan Recall yang tinggi menunjukkan bahwa model memiliki coverage yang baik (Recall tinggi) tetapi kurang akurat dalam menempatkan item relevan di peringkat teratas (Precision rendah).
Ini bisa menjadi indikasi bahwa model cenderung memberikan lebih banyak rekomendasi yang mungkin tidak relevan di posisi teratas, meskipun ia cukup baik dalam mengenali item yang disukai oleh pengguna secara keseluruhan.
keseluruhan.


#### Mencoba memberikan rekomendasi kepada user_id tertentu

In [37]:
user_id = 1.000030e+20 

if user_id in train_matrix.columns:
    print(f"{user_id} found in user_similarity.")
else:
    print(f"{user_id} not found in user_similarity.")

1.00003e+20 not found in user_similarity.


In [25]:
def get_recommendations_for_user(user_id, approx_matrix, original_matrix, n_recommendations=5):# Pastikan user_id adalah string jika diperlukan
    user_id = str(user_id)

    # Make sure the matrix index is a string
    approx_matrix.index = approx_matrix.index.astype(str)
    original_matrix.index = original_matrix.index.astype(str)

    # Check if user_id is in the original matrix index
    if user_id not in original_matrix.index:
        print(f"User ID {user_id} tidak ditemukan dalam data.")
        return None

    # Get rating prediction for a given user
    user_pred_ratings = approx_matrix.loc[user_id].values
    user_true_ratings = original_matrix.loc[user_id].values

    # Create a DataFrame from the original prediction and rating results.
    recommendations_df = pd.DataFrame({
        'item_id': original_matrix.columns,
        'predicted_rating': user_pred_ratings,
        'actual_rating': user_true_ratings
    })

    # Filter items that users haven't rated yet (recommend only new items)
    recommendations_df = recommendations_df[recommendations_df['actual_rating'] == 0]

    # Sort items by highest predicted rating
    recommendations_df = recommendations_df.sort_values(by='predicted_rating', ascending=False)

    # Take top-n recommendations
    top_recommendations = recommendations_df.head(n_recommendations)

    return top_recommendations

recommendations = get_recommendations_for_user(user_id, test_matrix_approx_df, test_matrix, n_recommendations=5)

# Displays recommendation results
if recommendations is not None:
    print(f"Top 5 rekomendasi untuk user {user_id}:")
    print(recommendations)

KeyError: '1.12603e+20'

Interpretasi Hasil Rekomendasi:
- Rekomendasi menunjukkan item yang belum pernah diberi rating oleh pengguna 1.12603e+20, dengan predicted_rating yang ditampilkan sebagai skor prediksi.
- Kolom actual_rating bernilai 0.0, yang menunjukkan bahwa pengguna ini belum memberi rating pada item-item tersebut sehingga item ini layak direkomendasikan.

Analisis Hasil:
1. Dalam hasil di atas, prediksi rating cukup rendah (di bawah 0.1). Ini bisa menunjukkan bahwa model SVD belum cukup efektif dalam memberikan rekomendasi yang kuat untuk pengguna ini. Hal ini mungkin terjadi jika data asli memiliki nilai rating yang jarang atau distribusi rating yang tidak merata, yang menyulitkan model untuk mempelajari pola yang kuat.
2. Rating prediksi yang rendah dapat mengindikasikan bahwa model kesulitan memberikan rekomendasi yang relevan bagi pengguna ini, atau pengguna tersebut memiliki sedikit interaksi sehingga model memiliki informasi yang terbatas tentang preferensinya.
3. Ini juga bisa menjadi tanda bahwa model SVD, meskipun baik untuk dekomposisi matriks, mungkin membutuhkan optimasi lebih lanjut atau perlu dikombinasikan dengan metode lain (seperti filtering berbasis konten atau deep learning) untuk meningkatkan kualitas rekomendasi.